src/checkpoint/dispatch/dispatch.impl.h

```diff
@@ -59,23 +64,125 @@ inline Serializer& operator|(Serializer& s, T& target) {
+
+  if (t.isPacking() || t.isSizing()) {
+    auto val = cleanType(&thisTypeIdx);
+    ap(t, val, len);
+  } else if (t.isUnpacking()) {
+    std::size_t serTypeIdx = 0;
+    auto val = cleanType(&serTypeIdx);
+    ap(t, val, len);
+
+    // Only temporary BEGIN
+    //
+
+    // try {
+    //   DoNothing<T>::doNothing();
+    // } catch (...) {
+    //   auto err =
+    //       std::string("Unpacking wrong type, got=") +
+    //       vrt::typeregistry::getTypeNameForIdx(thisTypeIdx) +
+    //       ", expected=" + vrt::typeregistry::getTypeNameForIdx(serTypeIdx);
+    //   throw std::runtime_error(err);
+    // }
+
+    //
+    // Only temporary END
+
+    if (vrt::typeregistry::validateIndex(serTypeIdx) == false ||
+        thisTypeIdx != serTypeIdx) {
+      auto err = std::string("Unpacking wrong type, got=") +
+                 vrt::typeregistry::getTypeNameForIdx(thisTypeIdx) +
+                 " idx=" + std::to_string(thisTypeIdx) + ", expected=" +
+                 vrt::typeregistry::getTypeNameForIdx(serTypeIdx) +
+                 " idx=" + std::to_string(serTypeIdx);
+      throw std::runtime_error(err);
+    }
+  }
+
+  return t;
+}
+
+template <typename T, typename TraverserT>
+TraverserT& Traverse::with(T& target, TraverserT& t, SerialSizeType len) {
   using CleanT = typename CleanType<T>::CleanT;
-  using DispatchType = typename TraverserT::template DispatcherType<TraverserT, CleanT>;
+  using DispatchType =
+      typename TraverserT::template DispatcherType<TraverserT, CleanT>;
 
-  auto val = cleanType(&target);
+#if defined SERIALIZATION_ERROR_CHECKING
+  withTypeIdx<CleanT>(t);
+#endif
 
+  auto val = cleanType(&target);
   SerializerDispatch<TraverserT, CleanT, DispatchType> ap;
+
+#if defined SERIALIZATION_ERROR_CHECKING
+  try {
+    ap(t, val, len);
+  } catch (std::runtime_error &err) {
+    auto const what = std::string{err.what()} +
+                      "\npath: " + vrt::typeregistry::getTypeName<T>();
+    throw std::runtime_error(what);
+  }
+#else
   ap(t, val, len);
+#endif
 
   return t;
 }
 
 template <typename T, typename TraverserT, typename... Args>
 TraverserT Traverse::with(T& target, Args&&... args) {
   TraverserT t(std::forward<Args>(args)...);
+
+#if !defined SERIALIZATION_ERROR_CHECKING
+  auto const thisTypeIdx = vrt::typeregistry::getTypeIdx<T>();
+  if (t.isPacking() || t.isSizing()) {
+    with(thisTypeIdx, t);
+  } else if (t.isUnpacking()) {
+    std::size_t serTypeIdx;
+    with(serTypeIdx, t);
+
+    if (vrt::typeregistry::validateIndex(serTypeIdx) == false ||
+        thisTypeIdx != serTypeIdx) {
+      auto err = std::string("Unpacking wrong type, got=") +
+                 vrt::typeregistry::getTypeNameForIdx(thisTypeIdx) +
+                 " idx=" + std::to_string(thisTypeIdx) + ", expected=" +
+                 vrt::typeregistry::getTypeNameForIdx(serTypeIdx) +
+                 " idx=" + std::to_string(serTypeIdx);
+      throw std::runtime_error(err);
+    }
+  }
+#endif
```

Comment:
That's the very basic check that only checks if `checkpoint::deserialize<T>(buffer)` is called for correct `T`. 
And there is no recursion for this check.

We wanted it to be enabled if "deep" checking is turned off.